### PR TITLE
Profile api endpoints code refactoring

### PR DESCRIPTION
**Pull request by swjeong0825**
Have two changes.

1. refactor the code (no logic change) for /referral/user endpoint
2. add the logic that restricts referral only to the jobPost creator.

### CODE REVIEW

File: backend/routes/referralRouters.ts [hunk 1]
```diff
@@ -42,22 +45,64 @@ const checkUserIsManager = (
...
+const checkOneJobPostIsCreatedByUser = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const jobPostId = req.query.jobId as string; //checkIfQueryHasValidJobId checks that it does exist.
+  const jobPostCreatorId = (
+    (await jobPostControllers.getOneJobPostWithId(jobPostId)) as JobPost
+  ).hiringManagerId; //checkIfQueryHasValidJobId checks that the jobPostId is the valid one.
+  if (req.user?.id === jobPostCreatorId) {
```
Comment (meagan-dalton) [inline backend/routes/referralRouters.ts:82]: I believe this line should say

if(req.user?.id !== jobPostCreatorId){ //logged in user does not match jobPostCreator so throw error
    next(new StatusCodedError("forbidden, not the jobpost creater", 403))
}

File: backend/routes/referralRouters.ts [hunk 2]
```diff
@@ -42,22 +45,64 @@ const checkUserIsManager = (
...
+  } else {
+    next();
+  }
+};
+
+const middlewaresForReferralJobPost = [
+  checkUserIsManager, //check if the user is manager
+  checkIfQueryHasValidJobId, //check if the req.query has valid jobId
+  checkOneJobPostIsCreatedByUser, //restrict to only the specific job creator and no one else
+];
```
Comment (meagan-dalton) [inline backend/routes/referralRouters.ts:93]: really cool, what you did here!

File: backend/routes/referralRouters.ts [hunk 3]
```diff
@@ -42,22 +45,64 @@ const checkUserIsManager = (
...
-referralRouter.get( //I feel like we should restrict this to only the specific job creator and no one else
-  "/jobs",
-  checkUserIsManager,
-  async (req: Request, res:Response, next: NextFunction) => {
-    try{
-      const jobId = req.query.jobId ? (req.query.jobId as string) : "";
-      const referrals = await referralControllers.getReferralsFromJobPostId(jobId);
-      res.status(200).json({referrals});
-    }catch(e:any){
+const checkIfQueryHasValidJobId = async (
```
Comment (meagan-dalton) [inline backend/routes/referralRouters.ts:55]: so smart, kind of forgot about throwing an error if job did not exist. good catch

**Review by meagan-dalton — CHANGES_REQUESTED** [review]
just the quick fix to line 60 of referralRouter and I think this will be all set to merge

File: backend/routes/referralRouters.ts (hunk 1 above)
Comment (swjeong0825) [reply backend/routes/referralRouters.ts:82]: It actually is. U can push the change

**Comment by swjeong0825** [comment]
> just the quick fix to line 60 of referralRouter and I think this will be all set to merge

Do u mean the jobpost.managerId !== req.user.id one? Or sth else?

**Pull request merged**